File: 程序员面试金典/4.4/isBalanced.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Result {
    int height;
    boolean isBalanced;

    Result(int height, boolean isBalanced) {
        this.height = height;
        this.isBalanced = isBalanced;
    }

}

class Solution {
    public boolean isBalanced(TreeNode root) {
        return isBalancedRec(root).isBalanced;
    }

    private Result isBalancedRec(TreeNode root) {
        if (root == null)
            return new Result(0, true);
        Result leftResult = isBalancedRec(root.left);
        Result rightResult = isBalancedRec(root.right);
        return new Result(
            1 + Math.max(leftResult.height, rightResult.height),
            leftResult.isBalanced && rightResult.isBalanced && Math.abs(leftResult.height - rightResult.height) <= 1
        );
    }
}

File: 剑指Offer/55-II/isBalanced.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        return balanced(root)[0]


def balanced(root):
    if not root:
        return True, 0
    lb, lh = balanced(root.left)
    rb, rh = balanced(root.right)
    return lb and rb and abs(lh - rh) <= 1, 1 + max(lh, rh)
